vignettes/invariants.Rmd

```diff
@@ -198,42 +192,147 @@ df[[TRUE]]
 df[[mean]]
 ```
 
-<!-- HW: the failure with mean needs a better error. -->
-
-Missing and out-of-bounds (OOB) values of `j` return `NULL` with a warning or error:
-
-```{r double-bracket-j-na, dftbl = TRUE}
-df[[NA_character_]]
-df[["x"]]
+Numeric NA, numeric out-of-bounds (OOB) values, and non-integers throw an error:
 
+```{r double-bracket-j-oob-numeric, dftbl = TRUE}
 df[[NA_integer_]]
 df[[-1]]
 df[[4]]
+df[[1.5]]
+df[[Inf]]
 ```
 
-<!-- HW: why is x[[NA]] an OOB error and not a type error? -->
+Character OOB access is silent because a common package idiom is to check for the absence of a column with `is.null(df[[var]])`.
 
-OOB access only generates a warning because a common existing idiom is to check for the absence of a column with `is.null(df$x)`.
+```{r double-bracket-j-oob-character, dftbl = TRUE}
+df[["x"]]
+```
 
-Non-integer numeric values produce an error:
+### Definition of `x$name`
 
-```{r, double-bracket-numeric, dftbl = TRUE}
-df[[1.5]]
+`x$name` is equal to `x[["name"]]`.
+Unlike data frames, tibbles do not partially match names.
+
+```{r dollar-equivalent-to-subset-pmatch, dftbl = TRUE}
+df$l
 ```
 
-## Definition of `x$name`
+Because `df$x` is rarely used in packages, it can raise a warning:
 
-`x$name` is equal to `x[["name"]]`.
+```{r dollar-j-oob-character, dftbl = TRUE}
+df$not_present
+```
 
-<!-- I removed the examples here because I think the principle is so simple -->
+## Column subsetting
 
-Unlike data frames, tibbles do not partially match.
+### Definition of `x[j]`.
 
-```{r dollar-equivalent-to-subset-pmatch, dftbl = TRUE}
-df$l
+`j` is converted to an integer vector by `vectrs::vec_as_index(j, nrow(x), names = names(x))`. Then `x[c(j_1, j_2, ..., j_n)]` is equivalent to `tibble(x[[j_1]], x[[j_2]], ..., x[[j_3]])``.
+This implies that `j` must be a numeric or character vector, or a logical vector with length 1 or `nrow(x)`.
+
+```{r bracket-j-definition, dftbl = TRUE}
+df[1:2]
 ```
 
```

Comment:
NB: `x[j][[jj]]` equal to `x[ j[[jj]] ][[1]]` ?

Reply:
`x[j][[jj]]` is equal to `x [[ j[[j]] ]]`